lib/async_hooks.js

```diff
@@ -342,24 +345,28 @@ function emitInitS(asyncId, type, triggerAsyncId, resource) {
   }
 }
 
-
-function emitBeforeN(asyncId) {
-  processing_hook = true;
-  // Use a single try/catch for all hook to avoid setting up one per iteration.
-  try {
-    for (var i = 0; i < active_hooks_array.length; i++) {
-      if (typeof active_hooks_array[i][before_symbol] === 'function') {
-        active_hooks_array[i][before_symbol](asyncId);
+function hookFactory(symbol) {
+  // Called from native. The asyncId stack handling is taken care of there
+  // before this is called.
+  return function(asyncId) {
```

Comment:
It does that on the newer v8 versions:

```js
> const a = { foo: function () {}, bar() {}, baz: () => {} }
undefined
> a.foo.name
'foo'
> a.bar.name
'bar'
> a.baz.name
'baz'
```

But I am not certain if there might be a context in which it is not inferred. I am also not sure where to trigger a stack trace from one of these functions as I didn't look deep into the async_hooks code so far. Can you point me out to something or give me a small example?

I could enforce setting the name though.

```js
function hookFactory(symbol, name) {
  const fn = function () {}
  Object.defineProperty(fn, 'name', {
    value: name
  })
  return fn
}

const hook = hookFactory(Symbol(), 'myName')
```